### ISSUE TITLE
`ActionContext.GetContext<string>("property")` returns json token of `{{ ValueKind: 3}}` instead of actual value.

### ISSUE THREAD

**Issue by ElanHasson** [microsoft/RulesEngine#179]
Hi,

I'm using 3.2.0.

I have a custom action in a action trigger:

```
 public class PropertySetterAction : ActionBase
    {
        private readonly RuleExpressionParser _expressionParser;

        public PropertySetterAction(RuleExpressionParser expressionParser, object input)
        {
            _expressionParser = expressionParser;
        }
        public override ValueTask<object> Run(ActionContext context, RuleParameter[] ruleParameters)
        {
            
            var propertyName = context.GetContext<string>("propertyName");
            var value = context.GetContext<string>("newValue");
            
            return new ValueTask<object>(this._expressionParser.Evaluate<object>(value, ruleParameters));
        }
    }
```
Debugger view: ![image](https://user-images.githubusercontent.com/234704/124035697-e398df80-d9ca-11eb-8c8f-ffd749d47612.png)

My Rules look like this:

```json
[
  {
    "WorkflowName": "Employee Validations",
    "Rules": [
      {
        "RuleName": "If Employee Last name is unset, use first name as last name",
        "Expression": "!string.IsNullOrWhiteSpace(input1.LastName)",
        "Actions": {
          "OnFailure": {
            "Name": "PropertySetterAction",
            "Context": {
              "propertyName": "LastName",
              "newValue": "input1.FirstName"
             }
          }
        }
      }
    ]
  }
]
```

My expectation is that `propertyName` in context would be the literal `LastName` and the value of `newValue` would be evaluated to an expression.

Am I missing something or is there a defect?

**Comment by abbasc52** [comment]
Hi @ElanHasson ,
I am unable to reproduce this error.
Can you please provide more details on the following so that we can help you further:
- Which dotnet version are you using?
- How are you parsing the json rule?

**Comment by ElanHasson** [comment]
Thanks for your response.
I forgot to press save on my update to the issue last night. 

I found when debugging the RulesEngine library it's a System.Text.Json thing as that's how I'm deserializing my rules json which is getting serialized in  at https://github.com/microsoft/RulesEngine/blob/main/src/RulesEngine/Actions/ActionContext.cs#L22

I suspect when deserializing, I should specify the `WorkflowRules` type versus allowing `dynamic` to be returned.

It may make sense for me to send a PR to add support for choosing a serializer to use.

Please let me know your thoughts.

**Comment by abbasc52** [comment]
@ElanHasson thanks for the reply. Currently the support for System.Text.Json is only partial.
Probably we can take it as a separate issue on improving System.Text.Json support. 

As for your issue, i believe using newtonsoft.json for json parsing should fix your issue.

**Comment by ElanHasson** [comment]
Thanks will update here.

**Comment by ElanHasson** [comment]
@abbasc52, what is the work necessary to support? I can send a PR.

Any interest in supporting whatever serializer via configurable extensibility points in `ReSettings`?

**Comment by abbasc52** [comment]
@ElanHasson I would need give some thought on this.
Currently we internally use newtonsoft only in 2 places:
1. When user passes workflows as array of strings
https://github.com/microsoft/RulesEngine/blob/bafbff281d16172bd7b4cf4c7785e684cfa2be2c/src/RulesEngine/RulesEngine.cs#L42-L46
2. When we try to parse context for actions
https://github.com/microsoft/RulesEngine/blob/65d2abbd8b28d631a7a886b2ad3760641423eb80/src/RulesEngine/Actions/ActionContext.cs#L22
https://github.com/microsoft/RulesEngine/blob/65d2abbd8b28d631a7a886b2ad3760641423eb80/src/RulesEngine/Actions/ActionContext.cs#L40

These are internal implementation, which should not impact users, so I need to think further whether we want to provide an ability to configure serializer or simply migrate to System.Text.Json

However, the issue with users trying to parse workflow json with System.Text.Json is that default serialization settings do not work well.
Mostly due to stringEnumConvertor.
https://github.com/microsoft/RulesEngine/blob/65d2abbd8b28d631a7a886b2ad3760641423eb80/src/RulesEngine/Models/Rule.cs#L42-L43

We have mentioned these attributes for newtonsoft, but not for System.Text.Json.


Also for actions, we have key as enum for `onSuccess` and `onFailure` which is not supported by System.Text.Json last time I checked.

https://github.com/microsoft/RulesEngine/blob/65d2abbd8b28d631a7a886b2ad3760641423eb80/src/RulesEngine/Models/Rule.cs#L48
 
From the looks of it enums are the main problem.
 
Now we can approach it two ways:
1. We can either change all enums to string and add proper validation on allowed values
2. Or add attributes for System.Text.Json as well on the model

I would like to know your opinion on this

**Comment by ElanHasson** [comment]
Thanks for the thorough response.

Judging from the overall direction of Json in .Net, there is a large investment on System.Text.Json in terms of perf. 

I have not been adding NewtonSoft.Json to any of my projects as direct dependencies for a while now.

I do not think serializer extensibility should be an objective here, it's a minor implementation detail. 

With that said, it may make sense to push workflow rule inputs  as strings to be a library consumer concern versus a function of the library and internally only work on the strongly- typed object model .


It would be nice to eliminate the json dependency as a whole.

The easiest solution is to add the attributes for now :)

**Comment by ElanHasson** [comment]
@abbasc52  I'm doing a PR to validate the change for the attributes for now. This will unblock me for now and we can figure out next steps.

**Comment by abbasc52** [comment]
@ElanHasson I digged deeper into the issue and realized that `System.Text.Json` does not play well with Object type and generates valuekind like object. 
https://josef.codes/custom-dictionary-string-object-jsonconverter-for-system-text-json/

It looks like something beyond the scope of this library, and we would recommend using Newtonsoft (or pass the json as string) till we are able to fully migrate to System.Text.Json

I will be closing this issue for now and will include this as a part of documentation

**Comment by ElanHasson** [comment]
No worries that's so much for what you've done.
